File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    jobs.sort()

    current_time = 0
    prev_time = -1
    
    total_time = 0 # answer = total_time / len(jobs)
    finish_cnt = 0 # 완료한 job 수
    wait_q = []
    
    while finish_cnt < len(jobs):
        for job in jobs:
            if prev_time < job[0] <= current_time:
                heapq.heappush(wait_q, (job[1],job[0])) # 소요시간, 요청시각 순
        
        if wait_q:
            running_job = heapq.heappop(wait_q)
            prev_time = current_time
            current_time += running_job[0]
            total_time += (current_time - running_job[1])
            finish_cnt += 1
            
        else:
            current_time+=1

    answer = int(total_time / len(jobs))
    return answer

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> jobs) {
    int n = jobs.size();
    int total_time = 0; // answer = total_time / jobs.size()
    int current_time = 0;
    int idx = 0; // jobs 순회
    
    // jobs 정렬
    sort(jobs.begin(), jobs.end());
    /*
    for (int i=0;i<n;i++){
        cout << jobs[i][0] << " " << jobs[i][1] << "\n";
    }
    */
    
    // jobs: {요청시각, 소요시간}
    // pq: {소요시간, 요청시각, 번호}
    priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
    
    while (idx < n || !pq.empty()){
        while (idx<n && jobs[idx][0] <= current_time){
            pq.push({jobs[idx][1], jobs[idx][0], idx});
            idx++;
        }
        
        if (!pq.empty()){
            current_time += pq.top()[0]; // +소요시간
            total_time += current_time - pq.top()[1]; // 완료시각 - 요청시각
            pq.pop();
        }
        else{
            current_time = jobs[idx][0];
        }
    }
    
    return total_time / n;
}